lib/presentation/blocs/activity_cubit.dart

```diff
@@ -0,0 +1,186 @@
+  ///
+  /// This only concerns memory-persisted location in a small scope and is used
+  /// for temporary loss of location such as if the user mistakenly disables
+  /// their location service. This does NOT corresponds to the last known
+  /// location from a previous session.
+  Location? _lastKnownLocation;
+
+  /// Listens to changes in the [PermissionsCubit].
+  ///
+  /// If a permission changes eventually retry to get the location if the
+  /// location stream was on error.
+  void _initPermissionsSubscription() {
+    _permissionsSubscription = _params.permissionsCubit?.stream.listen(
+      (permissionsState) {
+        // If the permissions state changes
+        permissionsState.mapOrNull(
+          loaded: (loadedPermissions) {
+            // Location permission is granted
+            if (loadedPermissions.locationPermission?.getOrNull()?.isGranted ??
+                false) {
+              // Current state is error, retry to get location as the cause for
+              // error has gone
+              state.mapOrNull(
+                error: (_) {
+                  retryTrackLocation();
+                },
+              );
+            }
+          },
+        );
+      },
+    );
+  }
+
+  /// Closes the location subscription and restarts it.
+  void retryTrackLocation() async {
+    await _closeLocationSubscription();
+    listenToLocation();
+  }
+
+  /// Starts a subscription to the user location using [GetLocationStream].
+  ///
+  /// Cancels any previous subscription before attempting a new one.
+  ///
+  /// If the location stream fails, [ActivityState.error] is emitted with
+  /// [_lastKnownLocation] set if available.
+  ///
+  /// Updates [_lastKnownLocation] each time a new location is emitted from the
+  /// stream.
+  void listenToLocation() {
+    emit(ActivityState.loading(lastKnownLocation: _lastKnownLocation));
+    _locationSubscriptionMutex.protect(
+      () async {
+        if (_locationSubscription != null) {
+          await _locationSubscription!.cancel();
+        }
+        _locationSubscription = _getLocationStream(
+          _params.notificationConfig,
+        ).listen(
+          (locationResult) {
+            locationResult.fold(
+              (timedLocation) {
+                _lastKnownLocation = timedLocation.location;
+                emit(
+                  ActivityState.loaded(
+                    currentLocation: timedLocation.location,
+                  ),
+                );
+              },
+              (fault) {
+                // If state is loaded ignore the failure, otherwise emit an
+                // error
+                state.maybeMap(
+                  loaded: (_) {},
+                  orElse: () {
+                    emit(
+                      ActivityState.error(
+                        fault: fault,
+                        lastKnownLocation: _lastKnownLocation,
+                      ),
+                    );
+                  },
+                );
```

Comment:
Why ignore faults when the state is loaded ?

Reply:
It was before I added the `lastKnownLocation` property to the loading and error states, I'll change that